// @sweet-js/core/dist/apply-scope-in-params-reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _terms = require("./terms");

var _terms2 = _interopRequireDefault(_terms);

var _symbol = require("./symbol");

var _transforms = require("./transforms");

var _errors = require("./errors");

var _syntax = require("./syntax");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ScopeApplyingReducer {
  constructor(scope, context) {
    this.context = context;
    this.scope = scope;
  }

  transform(term) {
    let field = "transform" + term.type;
    if (typeof this[field] === 'function') {
      return this[field](term);
    }
    (0, _errors.assert)(false, "transform not implemented yet for: " + term.type);
  }

  transformFormalParameters(term) {
    let rest = term.rest == null ? null : this.transform(term.rest);
    return new _terms2.default('FormalParameters', {
      items: term.items.map(it => this.transform(it)),
      rest: rest
    });
  }

  transformBindingWithDefault(term) {
    return new _terms2.default('BindingWithDefault', {
      binding: this.transform(term.binding),
      init: term.init
    });
  }

  transformObjectBinding(term) {
    // TODO: much more complicated logic here
    return term;
    // return new Term('ObjectBinding', {
    //   properties: term.properties.map(prop => this.transform(prop))
    // });
  }

  transformBindingPropertyIdentifier(term) {
    return new _terms2.default('BindingPropertyIdentifier', {
      binding: this.transform(term.binding),
      init: term.init
    });
  }

  transformBindingPropertyProperty(term) {
    return new _terms2.default('BindingPropertyProperty', {
      name: term.name,
      binding: this.transform(term.binding)
    });
  }

  transformArrayBinding(term) {
    return new _terms2.default('ArrayBinding', {
      elements: term.elements.map(el => this.transform(el)),
      restElement: term.restElement == null ? null : this.transform(term.restElement)
    });
  }

  transformBindingIdentifier(term) {
    let name = term.name.addScope(this.scope, this.context.bindings, _syntax.ALL_PHASES);
    let newBinding = (0, _symbol.gensym)(name.val());

    this.context.env.set(newBinding.toString(), new _transforms.VarBindingTransform(name));
    this.context.bindings.add(name, {
      binding: newBinding,
      phase: this.context.phase,
      skipDup: true
    });

    return new _terms2.default("BindingIdentifier", { name: name });
  }
}
exports.default = ScopeApplyingReducer;
