// @sweet-js/core/dist/compiler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _termExpander = require('./term-expander.js');

var _termExpander2 = _interopRequireDefault(_termExpander);

var _tokenExpander = require('./token-expander');

var _tokenExpander2 = _interopRequireDefault(_tokenExpander);

var _ramda = require('ramda');

var _ = _interopRequireWildcard(_ramda);

var _multimap = require('./multimap');

var _multimap2 = _interopRequireDefault(_multimap);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Compiler {
  constructor(phase, env, store, context) {
    this.phase = phase;
    this.env = env;
    this.store = store;
    this.invokedRegistry = new _multimap2.default();
    this.context = context;
  }

  compile(stxl) {
    let tokenExpander = new _tokenExpander2.default(_.merge(this.context, {
      phase: this.phase,
      env: this.env,
      store: this.store,
      invokedRegistry: this.invokedRegistry
    }));
    let termExpander = new _termExpander2.default(_.merge(this.context, {
      phase: this.phase,
      env: this.env,
      store: this.store,
      invokedRegistry: this.invokedRegistry
    }));

    return tokenExpander.expand(stxl).map(t => termExpander.expand(t));
  }
}
exports.default = Compiler;
