// @sweet-js/core/dist/load-syntax.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expandCompiletime = expandCompiletime;
exports.sanitizeReplacementValues = sanitizeReplacementValues;
exports.evalCompiletimeValue = evalCompiletimeValue;

var _sweetSpec = require('sweet-spec');

var S = _interopRequireWildcard(_sweetSpec);

var _ramda = require('ramda');

var _ = _interopRequireWildcard(_ramda);

var _immutable = require('immutable');

var _syntax = require('./syntax');

var _syntax2 = _interopRequireDefault(_syntax);

var _shiftCodegen = require('shift-codegen');

var _shiftCodegen2 = _interopRequireDefault(_shiftCodegen);

var _sweetToShiftReducer = require('./sweet-to-shift-reducer');

var _sweetToShiftReducer2 = _interopRequireDefault(_sweetToShiftReducer);

var _termExpander = require('./term-expander');

var _termExpander2 = _interopRequireDefault(_termExpander);

var _env = require('./env');

var _env2 = _interopRequireDefault(_env);

var _templateProcessor = require('./template-processor');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function expandCompiletime(term, context) {
  // each compiletime value needs to be expanded with a fresh
  // environment and in the next higher phase
  let syntaxExpander = new _termExpander2.default(_.merge(context, {
    phase: context.phase + 1,
    env: new _env2.default(),
    store: context.store
  }));

  return syntaxExpander.expand(term);
}

function sanitizeReplacementValues(values) {
  if (Array.isArray(values)) {
    return sanitizeReplacementValues((0, _immutable.List)(values));
  } else if (_immutable.List.isList(values)) {
    return values.map(sanitizeReplacementValues);
  } else if (values == null) {
    throw new Error('replacement values for syntax template must not be null or undefined');
  } else if (typeof values.next === 'function') {
    return sanitizeReplacementValues((0, _immutable.List)(values));
  }
  return values;
}

// (Expression, Context) -> [function]
function evalCompiletimeValue(expr, context) {
  let sandbox = {
    syntaxTemplate: function (ident, ...values) {
      return (0, _templateProcessor.replaceTemplate)(context.templateMap.get(ident), sanitizeReplacementValues(values));
    }
  };

  let sandboxKeys = (0, _immutable.List)(Object.keys(sandbox));
  let sandboxVals = sandboxKeys.map(k => sandbox[k]).toArray();

  let parsed = new S.Module({
    directives: (0, _immutable.List)(),
    items: _immutable.List.of(new S.ExpressionStatement({
      expression: new S.FunctionExpression({
        isGenerator: false,
        name: null,
        params: new S.FormalParameters({
          items: sandboxKeys.map(param => {
            return new S.BindingIdentifier({
              name: _syntax2.default.from('identifier', param)
            });
          }),
          rest: null
        }),
        body: new S.FunctionBody({
          directives: _immutable.List.of(new S.Directive({
            rawValue: 'use strict'
          })),
          statements: _immutable.List.of(new S.ReturnStatement({
            expression: expr
          }))
        })
      })
    }))
  }).reduce(new _sweetToShiftReducer2.default(context.phase));

  let gen = (0, _shiftCodegen2.default)(parsed, new _shiftCodegen.FormattedCodeGen());
  let result = context.transform(gen);

  let val = context.loader.eval(result.code, context.store);
  return val.apply(undefined, sandboxVals);
}
