// @sweet-js/core/dist/modules.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Modules = exports.Module = undefined;

var _immutable = require("immutable");

var _env = require("./env");

var _env2 = _interopRequireDefault(_env);

var _store = require("./store");

var _store2 = _interopRequireDefault(_store);

var _tokenReader = require("./reader/token-reader");

var _tokenReader2 = _interopRequireDefault(_tokenReader);

var _ramda = require("ramda");

var _ = _interopRequireWildcard(_ramda);

var _symbol = require("./symbol");

var _terms = require("./terms");

var T = _interopRequireWildcard(_terms);

var _loadSyntax = require("./load-syntax");

var _compiler = require("./compiler");

var _compiler2 = _interopRequireDefault(_compiler);

var _transforms = require("./transforms");

var _scope = require("./scope");

var _errors = require("./errors");

var _hygieneUtils = require("./hygiene-utils");

var _syntax = require("./syntax");

var _utilsDirname = require("utils-dirname");

var _utilsDirname2 = _interopRequireDefault(_utilsDirname);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Module {
  constructor(moduleSpecifier, isNative, importEntries, exportEntries, pragmas, body) {
    this.moduleSpecifier = moduleSpecifier;
    this.isNative = isNative;
    this.importEntries = importEntries;
    this.exportEntries = exportEntries;
    this.pragmas = pragmas;
    this.body = body;
  }
}

exports.Module = Module;
const findBindingIdentifierName = term => {
  // TODO: handle destructuring
  (0, _errors.assert)(term.name, `not implemented yet for type ${ term.type }`);
  return term.name;
};

const convertExport = term => {
  let declaration = term.declaration;
  let bindings = [];
  if (T.isVariableDeclaration(declaration)) {
    bindings = declaration.declarators.map(decl => findBindingIdentifierName(decl.binding));
  } else if (T.isFunctionDeclaration(declaration) || T.isClassDeclaration(declaration)) {
    bindings.push(findBindingIdentifierName(declaration.name));
  }

  let namedExports = bindings.map(binding => {
    return new T.default('ExportSpecifier', {
      name: null,
      exportedName: binding
    });
  });
  return new T.default('ExportFrom', {
    moduleSpecifier: null,
    namedExports: (0, _immutable.List)(namedExports)
  });
};

const pragmaRegep = /^\s*#\w*/;

class Modules {
  constructor(context) {
    this.compiledModules = new Map();
    this.context = context;
    this.context.modules = this;
  }

  loadString(str) {
    let checkPragma = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

    let hasPragma = pragmaRegep.test(str);
    if (checkPragma && !hasPragma) {
      return {
        isNative: true,
        body: (0, _immutable.List)()
      };
    }
    const result = {
      isNative: !hasPragma,
      body: (0, _tokenReader2.default)(str)
    };
    return result;
  }

  load(path) {
    // TODO resolve and we need to carry the cwd through correctly
    return this.loadString(this.context.moduleLoader(path));
  }

  compile(mod, path) {
    let stxl = mod.body;
    let outScope = (0, _scope.freshScope)('outsideEdge');
    let inScope = (0, _scope.freshScope)(`insideEdge0`);
    // the compiler starts at phase 0, with an empty environment and store
    let compiler = new _compiler2.default(0, new _env2.default(), new _store2.default(), _.merge(this.context, {
      currentScope: [outScope, inScope],
      cwd: path === '<<entrypoint>>' ? this.context.cwd : (0, _utilsDirname2.default)(path)
    }));
    let terms = compiler.compile(stxl.map(s => s.addScope(outScope, this.context.bindings, _syntax.ALL_PHASES).addScope(inScope, this.context.bindings, 0)));

    let importEntries = [];
    let exportEntries = [];
    let pragmas = [];
    let filteredTerms = terms.reduce((acc, t) => {
      return _.cond([[T.isImport, t => {
        importEntries.push(t);
        return acc;
      }], [T.isExport, t => {
        // exportEntries.push(t);
        // return acc.concat(t);
        if (t.declaration) {
          exportEntries.push(convertExport(t));
          if (T.isVariableDeclaration(t.declaration)) {
            return acc.concat(new T.default('VariableDeclarationStatement', {
              declaration: t.declaration
            }));
          }
          return acc.concat(t.declaration);
        }
        exportEntries.push(t);
        return acc;
      }], [T.isPragma, t => {
        pragmas.push(t);return acc;
      }], [_.T, t => acc.concat(t)]])(t);
    }, (0, _immutable.List)());
    return new Module(path, mod.isNative, (0, _immutable.List)(importEntries), (0, _immutable.List)(exportEntries), (0, _immutable.List)(pragmas), filteredTerms);
  }

  compileEntrypoint(source, filename) {
    let enforcePragma = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

    let stxl = this.loadString(source, false);
    if (enforcePragma && stxl.isNative) {
      throw new Error(`Entrypoint ${ filename } must begin with #lang pragma`);
    }
    return this.getAtPhase('<<entrypoint>>', 0, this.context.cwd, stxl);
  }

  // Modules have a unique scope per-phase. We compile each module once at
  // phase 0 and store the compiled module in a map. Then, as we ask for
  // the module in a particular phase, we add that new phase-specific scope
  // to the compiled module and update the map with the module at that specific
  // phase.
  getAtPhase(rawPath, phase, cwd) {
    let rawStxl = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;

    let path = rawPath === '<<entrypoint>>' ? rawPath : this.context.moduleResolver(rawPath, cwd);
    let mapKey = `${ path }:${ phase }`;
    if (!this.compiledModules.has(mapKey)) {
      if (phase === 0) {
        let stxl = rawStxl != null ? rawStxl : this.load(path);
        this.compiledModules.set(mapKey, this.compile(stxl, path));
      } else {
        let rawMod = this.getAtPhase(rawPath, 0, cwd, rawStxl);
        let scope = (0, _scope.freshScope)(`insideEdge${ phase }`);
        this.compiledModules.set(mapKey, new Module(rawMod.moduleSpecifier, false, rawMod.importEntries.map(term => term.addScope(scope, this.context.bindings, phase)), rawMod.exportEntries.map(term => term.addScope(scope, this.context.bindings, phase)), rawMod.pragmas, rawMod.body.map(term => term.addScope(scope, this.context.bindings, phase))));
      }
    }
    return this.compiledModules.get(mapKey);
  }

  has(rawPath) {
    let phase = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;

    let path = rawPath === '<<entrypoint>>' ? rawPath : this.context.moduleResolver(rawPath, this.context.cwd);
    let key = `${ path }:${ phase }`;
    return this.compiledModules.has(key) && !this.compiledModules.get(key).isNative;
  }

  registerSyntaxDeclaration(term, phase, store) {
    term.declarators.forEach(decl => {
      let val = (0, _loadSyntax.evalCompiletimeValue)(decl.init.gen(), _.merge(this.context, {
        phase: phase + 1, store: store
      }));
      (0, _hygieneUtils.collectBindings)(decl.binding).forEach(stx => {
        if (phase !== 0) {
          // phase 0 bindings extend the binding map during compilation
          let newBinding = (0, _symbol.gensym)(stx.val());
          this.context.bindings.add(stx, {
            binding: newBinding,
            phase: phase,
            skipDup: false
          });
        }
        let resolvedName = stx.resolve(phase);
        store.set(resolvedName, new _transforms.CompiletimeTransform(val));
      });
    });
  }

  registerVariableDeclaration(term, phase, store) {
    term.declarators.forEach(decl => {
      (0, _hygieneUtils.collectBindings)(decl.binding).forEach(stx => {
        if (phase !== 0) {
          // phase 0 bindings extend the binding map during compilation
          let newBinding = (0, _symbol.gensym)(stx.val());
          this.context.bindings.add(stx, {
            binding: newBinding,
            phase: phase,
            skipDup: term.kind === 'var'
          });
        }
        let resolvedName = stx.resolve(phase);
        store.set(resolvedName, new _transforms.VarBindingTransform(stx));
      });
    });
  }

  registerFunctionOrClass(term, phase, store) {
    (0, _hygieneUtils.collectBindings)(term.name).forEach(stx => {
      if (phase !== 0) {
        let newBinding = (0, _symbol.gensym)(stx.val());
        this.context.bindings.add(stx, {
          binding: newBinding,
          phase: phase,
          skipDup: false
        });
      }
      let resolvedName = stx.resolve(phase);
      store.set(resolvedName, new _transforms.VarBindingTransform(stx));
    });
  }

  visit(mod, phase, store) {
    // TODO: recursively visit imports
    mod.body.forEach(term => {
      if (T.isSyntaxDeclarationStatement(term)) {
        this.registerSyntaxDeclaration(term.declaration, phase, store);
      }
    });
    return store;
  }

  invoke(mod, phase, store) {
    // TODO: recursively visit imports
    let body = mod.body.filter(_.complement(T.isCompiletimeStatement)).map(term => {
      term = term.gen(); // TODO: can we remove the need for gen? have to deeply remove compiletime code
      if (T.isVariableDeclarationStatement(term)) {
        this.registerVariableDeclaration(term.declaration, phase, store);
      } else if (T.isFunctionDeclaration(term)) {
        this.registerFunctionOrClass(term, phase, store);
      }
      return term;
    });
    (0, _loadSyntax.evalRuntimeValues)(body, _.merge(this.context, {
      store: store, phase: phase
    }));
    return store;
  }
}
exports.Modules = Modules;
