// @sweet-js/core/dist/module-visitor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isBoundToCompiletime = isBoundToCompiletime;
exports.bindImports = bindImports;

var _loadSyntax = require('./load-syntax');

var _ramda = require('ramda');

var _ = _interopRequireWildcard(_ramda);

var _sweetSpec = require('sweet-spec');

var T = _interopRequireWildcard(_sweetSpec);

var _sweetSpecUtils = require('./sweet-spec-utils');

var S = _interopRequireWildcard(_sweetSpecUtils);

var _symbol = require('./symbol');

var _transforms = require('./transforms');

var _hygieneUtils = require('./hygiene-utils');

var _sweetModule = require('./sweet-module');

var _sweetModule2 = _interopRequireDefault(_sweetModule);

var _immutable = require('immutable');

var _sweetToShiftReducer = require('./sweet-to-shift-reducer');

var _sweetToShiftReducer2 = _interopRequireDefault(_sweetToShiftReducer);

var _shiftCodegen = require('shift-codegen');

var _shiftCodegen2 = _interopRequireDefault(_shiftCodegen);

var _syntax = require('./syntax');

var _syntax2 = _interopRequireDefault(_syntax);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function isBoundToCompiletime(name, store) {
  let resolvedName = name.resolve(0);
  if (store.has(resolvedName)) {
    return store.get(resolvedName) instanceof _transforms.CompiletimeTransform;
  }
  return false;
}

function bindImports(impTerm, exModule, phase, context, isEntrypoint) {
  let names = [];
  let phaseToBind = impTerm.forSyntax ? phase + 1 : phase;
  if (impTerm.defaultBinding != null && impTerm instanceof T.Import) {
    let name = impTerm.defaultBinding.name;
    let exportName = exModule.exportedNames.find(exName => exName.exportedName.val() === '_default');
    if (exportName != null) {
      let newBinding = (0, _symbol.gensym)('_default');
      let toForward = exportName.exportedName;

      if (!isEntrypoint || isBoundToCompiletime(toForward, context.store) || impTerm.forSyntax) {
        context.bindings.addForward(name, toForward, newBinding, phaseToBind);
      }
      names.push(name);
    }
  }
  if (impTerm.namedImports) {
    impTerm.namedImports.forEach(specifier => {
      let name = specifier.binding.name;
      let exportName = exModule.exportedNames.find(exName => exName.exportedName.val() === name.val());
      if (exportName != null) {
        let newBinding = (0, _symbol.gensym)(name.val());
        let toForward = exportName.name ? exportName.name : exportName.exportedName;
        if (!isEntrypoint || isBoundToCompiletime(toForward, context.store) || impTerm.forSyntax) {
          context.bindings.addForward(name, toForward, newBinding, phaseToBind);
        }
        names.push(name);
      }
    });
  }
  if (impTerm.namespaceBinding) {
    let name = impTerm.namespaceBinding.name;
    let newBinding = (0, _symbol.gensym)(name.val());
    context.store.set(newBinding.toString(), new _transforms.ModuleNamespaceTransform(name, exModule));
    context.bindings.add(name, {
      binding: newBinding,
      phase: phaseToBind,
      skipDup: false
    });

    names.push(name);
  }
  return (0, _immutable.List)(names);
}

exports.default = class {

  constructor(context) {
    this.context = context;
  }

  visit(mod, phase, store, cwd) {
    mod.imports.forEach(imp => {
      if (imp.forSyntax) {
        let mod = this.context.loader.get(imp.moduleSpecifier.val(), phase + 1, cwd);
        this.visit(mod, phase + 1, store, mod.path);
        this.invoke(mod, phase + 1, store, mod.path);
      } else {
        let mod = this.context.loader.get(imp.moduleSpecifier.val(), phase, cwd);
        this.visit(mod, phase, store, mod.path);
      }
      bindImports(imp, mod, phase, this.context, false);
    });
    for (let term of mod.compiletimeItems()) {
      if (S.isSyntaxDeclarationStatement(term)) {
        this.registerSyntaxDeclaration(term.declaration, phase, store);
      }
    }
    return store;
  }

  invoke(mod, phase, store, cwd) {
    if (this.context.invokedRegistry.containsAt(mod.path, phase)) {
      return store;
    }
    mod.imports.forEach(imp => {
      if (!imp.forSyntax) {
        let mod = this.context.loader.get(imp.moduleSpecifier.val(), phase, cwd);
        this.invoke(mod, phase, store, mod.path);
        bindImports(imp, mod, phase, this.context, false);
      }
    });
    let items = mod.runtimeItems();
    for (let term of items) {
      if (S.isVariableDeclarationStatement(term)) {
        this.registerVariableDeclaration(term.declaration, phase, store);
      } else if (S.isFunctionDeclaration(term)) {
        this.registerFunctionOrClass(term, phase, store);
      }
    }
    let parsed = new T.Module({
      directives: (0, _immutable.List)(),
      items
      // $FlowFixMe: flow doesn't know about reduce yet
    }).reduce(new _sweetToShiftReducer2.default(phase));

    let gen = (0, _shiftCodegen2.default)(parsed, new _shiftCodegen.FormattedCodeGen());
    let result = this.context.transform(gen);

    this.context.loader.eval(result.code, store);
    this.context.invokedRegistry.add(mod.path, phase);
    return store;
  }

  registerSyntaxDeclaration(term, phase, store) {
    term.declarators.forEach(decl => {
      let val = (0, _loadSyntax.evalCompiletimeValue)(decl.init, _.merge(this.context, {
        phase: phase + 1,
        store
      }));

      (0, _hygieneUtils.collectBindings)(decl.binding).forEach(stx => {
        if (phase !== 0) {
          // phase 0 bindings extend the binding map during compilation
          let newBinding = (0, _symbol.gensym)(stx.val());
          this.context.bindings.add(stx, {
            binding: newBinding,
            phase: phase,
            skipDup: false
          });
        }
        let resolvedName = stx.resolve(phase);
        let compiletimeType = term.kind === 'operator' ? 'operator' : 'syntax';
        store.set(resolvedName, new _transforms.CompiletimeTransform({
          type: compiletimeType,
          prec: decl.prec == null ? void 0 : decl.prec.val(),
          assoc: decl.assoc == null ? void 0 : decl.assoc.val(),
          f: val
        }));
      });
    });
  }

  registerVariableDeclaration(term, phase, store) {
    term.declarators.forEach(decl => {
      (0, _hygieneUtils.collectBindings)(decl.binding).forEach(stx => {
        if (phase !== 0) {
          // phase 0 bindings extend the binding map during compilation
          let newBinding = (0, _symbol.gensym)(stx.val());
          this.context.bindings.add(stx, {
            binding: newBinding,
            phase: phase,
            skipDup: term.kind === 'var'
          });
        }
      });
    });
  }

  registerFunctionOrClass(term, phase, store) {
    (0, _hygieneUtils.collectBindings)(term.name).forEach(stx => {
      if (phase !== 0) {
        let newBinding = (0, _symbol.gensym)(stx.val());
        this.context.bindings.add(stx, {
          binding: newBinding,
          phase: phase,
          skipDup: false
        });
      }
    });
  }
};
