// @sweet-js/core/dist/node-loader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _sweetLoader = require('./sweet-loader');

var _sweetLoader2 = _interopRequireDefault(_sweetLoader);

var _fs = require('fs');

var _path = require('path');

var _resolve = require('resolve');

var _resolve2 = _interopRequireDefault(_resolve);

var _vm = require('vm');

var _vm2 = _interopRequireDefault(_vm);

var _store = require('./store');

var _store2 = _interopRequireDefault(_store);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class NodeLoader extends _sweetLoader2.default {

  constructor(baseDir, options = {}) {
    super(baseDir, options);
    this.extensions = options.extensions || ['.js', '.mjs'];
  }

  normalize(name, refererName, refererAddress) {
    let normName = super.normalize(name, refererName, refererAddress);
    let match = normName.match(_sweetLoader.phaseInModulePathRegexp);
    if (match && match.length >= 3) {
      let resolvedName = match[1];
      try {
        resolvedName = _resolve2.default.sync(match[1], {
          basedir: refererName ? (0, _path.dirname)(refererName) : this.baseDir,
          extensions: this.extensions
        });
      } catch (e) {
        // ignored
      }
      return `${resolvedName}:${match[2]}`;
    }
    throw new Error(`Module ${name} is missing phase information`);
  }

  fetch({
    name,
    address,
    metadata
  }) {
    let src = this.sourceCache.get(address.path);
    if (src != null) {
      return src;
    } else {
      try {
        src = (0, _fs.readFileSync)(address.path, 'utf8');
      } catch (e) {
        src = '';
      }
      this.sourceCache.set(address.path, src);
      return src;
    }
  }

  freshStore() {
    let sandbox = {
      process: global.process,
      console: global.console
    };
    return new _store2.default(_vm2.default.createContext(sandbox));
  }

  eval(source, store) {
    return _vm2.default.runInContext(source, store.getBackingObject());
  }
}
exports.default = NodeLoader;
