// @sweet-js/core/dist/scope-reducer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _sweetSpec = require('sweet-spec');

var S = _interopRequireWildcard(_sweetSpec);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

// $FlowFixMe: flow doesn't know about the CloneReducer yet
exports.default = class extends S.default.CloneReducer {

  constructor(scopes, bindings) {
    super();
    this.scopes = scopes;
    this.bindings = bindings;
  }

  applyScopes(s) {
    return this.scopes.reduce((acc, sc) => {
      return acc.addScope(sc.scope, this.bindings, sc.phase, {
        flip: sc.flip
      });
    }, s);
  }

  reduceBindingIdentifier(t, s) {
    return new S.BindingIdentifier({
      name: this.applyScopes(s.name)
    });
  }

  reduceIdentifierExpression(t, s) {
    return new S.IdentifierExpression({
      name: this.applyScopes(s.name)
    });
  }

  reduceRawSyntax(t, s) {
    // TODO: fix this once reading tokens is reasonable
    if (s.value.isTemplate() && s.value.items) {
      s.value.token.items = s.value.token.items.map(t => {
        if (t instanceof S.default) {
          return t.reduce(this);
        }
        return t;
      });
    }
    return new S.RawSyntax({
      value: this.applyScopes(s.value)
    });
  }
};
