// @sweet-js/core/dist/sweet-loader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.phaseInModulePathRegexp = undefined;

var _tokenReader = require('./reader/token-reader');

var _tokenReader2 = _interopRequireDefault(_tokenReader);

var _scope = require('./scope');

var _env = require('./env');

var _env2 = _interopRequireDefault(_env);

var _immutable = require('immutable');

var _compiler = require('./compiler');

var _compiler2 = _interopRequireDefault(_compiler);

var _syntax = require('./syntax');

var _bindingMap = require('./binding-map.js');

var _bindingMap2 = _interopRequireDefault(_bindingMap);

var _sweetSpec = require('sweet-spec');

var _sweetSpec2 = _interopRequireDefault(_sweetSpec);

var _sweetModule = require('./sweet-module');

var _sweetModule2 = _interopRequireDefault(_sweetModule);

var _ramda = require('ramda');

var _ = _interopRequireWildcard(_ramda);

var _scopeReducer = require('./scope-reducer');

var _scopeReducer2 = _interopRequireDefault(_scopeReducer);

var _macroContext = require('./macro-context');

var _babelCore = require('babel-core');

var _store = require('./store');

var _store2 = _interopRequireDefault(_store);

var _multimap = require('./multimap');

var _multimap2 = _interopRequireDefault(_multimap);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const phaseInModulePathRegexp = exports.phaseInModulePathRegexp = /(.*):(\d+)\s*$/;
class SweetLoader {

  constructor(baseDir, options = {}) {
    this.sourceCache = new Map();
    this.compiledCache = new Map();
    this.baseDir = baseDir;
    this.logging = options.logging || false;

    let bindings = new _bindingMap2.default();
    let templateMap = new Map();
    let tempIdent = 0;
    this.context = {
      phase: 0,
      bindings,
      templateMap,
      getTemplateIdentifier: () => ++tempIdent,
      loader: this,
      invokedRegistry: new _multimap2.default(),
      transform: c => {
        if (options.noBabel) {
          return {
            code: c
          };
        }
        return (0, _babelCore.transform)(c, {
          babelrc: true
        });
      }
    };
  }

  normalize(name, refererName, refererAddress) {
    // takes `..path/to/source.js:<phase>`
    // gives `/abs/path/to/source.js:<phase>`
    // missing phases are turned into 0
    if (!phaseInModulePathRegexp.test(name)) {
      return `${name}:0`;
    }
    return name;
  }

  locate({ name, metadata }) {
    // takes `/abs/path/to/source.js:<phase>`
    // gives { path: '/abs/path/to/source.js', phase: <phase> }
    let match = name.match(phaseInModulePathRegexp);
    if (match && match.length >= 3) {
      return {
        path: match[1],
        phase: parseInt(match[2], 10)
      };
    }
    throw new Error(`Module ${name} is missing phase information`);
  }

  fetch({
    name,
    address,
    metadata
  }) {
    throw new Error('No default fetch defined');
  }

  translate({
    name,
    address,
    source,
    metadata
  }) {
    let src = this.compiledCache.get(address.path);
    if (src != null) {
      return src;
    }
    let compiledModule = this.compileSource(source, address.path, metadata);
    this.compiledCache.set(address.path, compiledModule);
    return compiledModule;
  }

  instantiate({
    name,
    address,
    source,
    metadata
  }) {
    throw new Error('Not implemented yet');
  }

  eval(source) {
    return (0, eval)(source);
  }

  load(entryPath) {
    let metadata = {};
    let name = this.normalize(entryPath);
    let address = this.locate({ name, metadata });
    let source = this.fetch({ name, address, metadata });
    source = this.translate({ name, address, source, metadata });
    return this.instantiate({ name, address, source, metadata });
  }

  // skip instantiate
  compile(entryPath, {
    refererName,
    enforceLangPragma,
    isEntrypoint
  }) {
    let metadata = {
      isEntrypoint,
      enforceLangPragma,
      entryPath
    };
    let name = this.normalize(entryPath, refererName);
    let address = this.locate({ name, metadata });
    let source = this.fetch({ name, address, metadata });
    return this.translate({ name, address, source, metadata });
  }

  get(entryPath, entryPhase, refererName) {
    return this.compile(`${entryPath}:${entryPhase}`, {
      refererName,
      enforceLangPragma: true,
      isEntrypoint: false
    });
  }

  read(source) {
    return (0, _macroContext.wrapInTerms)((0, _tokenReader2.default)(source));
  }

  freshStore() {
    return new _store2.default({});
  }

  compileSource(source, path, metadata) {
    let directive = getLangDirective(source);
    if (directive == null && metadata.enforceLangPragma) {
      // eslint-disable-next-line no-console
      if (this.logging) console.log(`skipping module ${metadata.entryPath}`);
      return new _sweetModule2.default(path, _immutable.List.of());
    }
    let stxl = this.read(source);
    let outScope = (0, _scope.freshScope)('outsideEdge');
    let inScope = (0, _scope.freshScope)('insideEdge0');
    // the compiler starts at phase 0, with an empty environment and store
    let compiler = new _compiler2.default(0, new _env2.default(), this.freshStore(), _.merge(this.context, {
      currentScope: [outScope, inScope],
      cwd: path,
      isEntrypoint: metadata.isEntrypoint
    }));
    return new _sweetModule2.default(path, compiler.compile(stxl.map(s =>
    // $FlowFixMe: flow doesn't know about reduce yet
    s.reduce(new _scopeReducer2.default([{ scope: outScope, phase: _syntax.ALL_PHASES, flip: false }, { scope: inScope, phase: 0, flip: false }], this.context.bindings)))));
  }
}

exports.default = SweetLoader;
const langDirectiveRegexp = /\s*('lang .*')/;
function getLangDirective(source) {
  let match = source.match(langDirectiveRegexp);
  if (match) {
    return match[1];
  }
  return null;
}
