// @sweet-js/core/dist/sweet-spec-utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isClassDeclaration = exports.isFunctionDeclaration = exports.isCompiletimeStatement = exports.isSyntaxDeclarationStatement = exports.isVariableDeclarationStatement = exports.isSyntaxVariableDeclartion = exports.isVariableDeclarator = exports.isVariableDeclaration = exports.isExportFrom = exports.isExportDefault = exports.isExport = exports.isExportDeclaration = exports.isImportDeclaration = undefined;

var _sweetSpec = require('sweet-spec');

var T = _interopRequireWildcard(_sweetSpec);

var _ramda = require('ramda');

var _ = _interopRequireWildcard(_ramda);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

const isImportDeclaration = exports.isImportDeclaration = _.is(T.ImportDeclaration);

const isExportDeclaration = exports.isExportDeclaration = _.is(T.ExportDeclaration);
const isExport = exports.isExport = _.is(T.Export);
const isExportDefault = exports.isExportDefault = _.is(T.ExportDefault);
const isExportFrom = exports.isExportFrom = _.is(T.ExportFrom);

const isVariableDeclaration = exports.isVariableDeclaration = _.is(T.VariableDeclaration);
const isVariableDeclarator = exports.isVariableDeclarator = _.is(T.VariableDeclarator);
const isSyntaxVariableDeclartion = exports.isSyntaxVariableDeclartion = _.both(isVariableDeclaration, _.either(_.propEq('kind', 'syntax'), _.propEq('kind', 'syntaxrec')));

const isVariableDeclarationStatement = exports.isVariableDeclarationStatement = _.is(T.VariableDeclarationStatement);
const isSyntaxDeclarationStatement = exports.isSyntaxDeclarationStatement = term => {
  // syntax m = ...
  // syntaxrec m = ...
  return isVariableDeclarationStatement(term) && term.declaration.type === 'VariableDeclaration' && (term.declaration.kind === 'syntax' || term.declaration.kind === 'syntaxrec' || term.declaration.kind === 'operator');
};

const isCompiletimeStatement = exports.isCompiletimeStatement = isSyntaxDeclarationStatement;

const isFunctionDeclaration = exports.isFunctionDeclaration = _.is(T.FunctionDeclaration);
const isClassDeclaration = exports.isClassDeclaration = _.is(T.ClassDeclaration);
