// @sweet-js/core/dist/syntax.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ALL_PHASES = exports.Types = undefined;

var _immutable = require('immutable');

var _errors = require('./errors');

var _bindingMap = require('./binding-map');

var _bindingMap2 = _interopRequireDefault(_bindingMap);

var _ramdaFantasy = require('ramda-fantasy');

var _ramda = require('ramda');

var _ = _interopRequireWildcard(_ramda);

var _sweetSpec = require('sweet-spec');

var T = _interopRequireWildcard(_sweetSpec);

var _tokens = require('./tokens');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getFirstSlice(stx) {
  if (!stx || typeof stx.isDelimiter !== 'function') return null; // TODO: should not have to do this
  if (!stx.isDelimiter()) {
    return stx.token.slice;
  }
  return stx.token.get(0).token.slice;
}


function sizeDecending(a, b) {
  if (a.scopes.size > b.scopes.size) {
    return -1;
  } else if (b.scopes.size > a.scopes.size) {
    return 1;
  } else {
    return 0;
  }
}

let Types = exports.Types = {
  null: {
    match: token => !Types.delimiter.match(token) && token.type === _tokens.TokenType.NULL,
    create: (value, stx) => new Syntax({
      type: _tokens.TokenType.NULL,
      value: null,
      typeCode: _tokens.TypeCodes.Keyword
    }, stx)
  },
  number: {
    match: token => !Types.delimiter.match(token) && token.type.klass === _tokens.TokenClass.NumericLiteral,
    create: (value, stx) => new Syntax({
      type: _tokens.TokenType.NUMBER,
      value,
      typeCode: _tokens.TypeCodes.NumericLiteral
    }, stx)
  },
  string: {
    match: token => !Types.delimiter.match(token) && token.type.klass === _tokens.TokenClass.StringLiteral,
    create: (value, stx) => new Syntax({
      type: _tokens.TokenType.STRING,
      str: value,
      typeCode: _tokens.TypeCodes.StringLiteral
    }, stx)
  },
  punctuator: {
    match: token => !Types.delimiter.match(token) && token.type.klass === _tokens.TokenClass.Punctuator,
    create: (value, stx) => new Syntax({
      type: {
        klass: _tokens.TokenClass.Punctuator,
        name: value
      },
      typeCode: _tokens.TypeCodes.Punctuator,
      value
    }, stx)
  },
  keyword: {
    match: token => !Types.delimiter.match(token) && token.type.klass === _tokens.TokenClass.Keyword,
    create: (value, stx) => new Syntax({
      type: {
        klass: _tokens.TokenClass.Keyword,
        name: value
      },
      typeCode: _tokens.TypeCodes.Keyword,
      value
    }, stx)
  },
  identifier: {
    match: token => !Types.delimiter.match(token) && token.type.klass === _tokens.TokenClass.Ident,
    create: (value, stx) => new Syntax({
      type: _tokens.TokenType.IDENTIFIER,
      value,
      typeCode: _tokens.TypeCodes.Identifier
    }, stx)
  },
  regularExpression: {
    match: token => !Types.delimiter.match(token) && token.type.klass === _tokens.TokenClass.RegularExpression,
    create: (value, stx) => new Syntax({
      type: _tokens.TokenType.REGEXP,
      value,
      typeCode: _tokens.TypeCodes.RegExp
    }, stx)
  },
  // $FlowFixMe: cleanup all this
  braces: {
    match: token => Types.delimiter.match(token) && token.get(0).token.type === _tokens.TokenType.LBRACE,
    create: (inner, stx) => {
      let left = new T.RawSyntax({
        value: new Syntax({
          type: _tokens.TokenType.LBRACE,
          typeCode: _tokens.TypeCodes.Punctuator,
          value: '{',
          slice: getFirstSlice(stx)
        })
      });
      let right = new T.RawSyntax({
        value: new Syntax({
          type: _tokens.TokenType.RBRACE,
          typeCode: _tokens.TypeCodes.Punctuator,
          value: '}',
          slice: getFirstSlice(stx)
        })
      });
      return new T.RawDelimiter({
        kind: 'braces',
        inner: _immutable.List.of(left).concat(inner).push(right)
      });
    }
  },
  // $FlowFixMe: cleanup all this
  brackets: {
    match: token => Types.delimiter.match(token) && token.get(0).token.type === _tokens.TokenType.LBRACK,
    create: (inner, stx) => {
      let left = new T.RawSyntax({
        value: new Syntax({
          type: _tokens.TokenType.LBRACK,
          typeCode: _tokens.TypeCodes.Punctuator,
          value: '[',
          slice: getFirstSlice(stx)
        })
      });
      let right = new T.RawSyntax({
        value: new Syntax({
          type: _tokens.TokenType.RBRACK,
          typeCode: _tokens.TypeCodes.Punctuator,
          value: ']',
          slice: getFirstSlice(stx)
        })
      });
      return new T.RawDelimiter({
        kind: 'brackets',
        inner: _immutable.List.of(left).concat(inner).push(right)
      });
    }
  },
  // $FlowFixMe: cleanup all this
  parens: {
    match: token => Types.delimiter.match(token) && token.get(0).token.type === _tokens.TokenType.LPAREN,
    create: (inner, stx) => {
      let left = new T.RawSyntax({
        value: new Syntax({
          type: _tokens.TokenType.LPAREN,
          typeCode: _tokens.TypeCodes.Punctuator,
          value: '(',
          slice: getFirstSlice(stx)
        })
      });
      let right = new T.RawSyntax({
        value: new Syntax({
          type: _tokens.TokenType.RPAREN,
          typeCode: _tokens.TypeCodes.Punctuator,
          value: ')',
          slice: getFirstSlice(stx)
        })
      });
      return new T.RawDelimiter({
        kind: 'parens',
        inner: _immutable.List.of(left).concat(inner).push(right)
      });
    }
  },

  assign: {
    match: token => {
      if (Types.punctuator.match(token)) {
        switch (token.value) {
          case '=':
          case '|=':
          case '^=':
          case '&=':
          case '<<=':
          case '>>=':
          case '>>>=':
          case '+=':
          case '-=':
          case '*=':
          case '/=':
          case '%=':
            return true;
          default:
            return false;
        }
      }
      return false;
    }
  },

  boolean: {
    match: token => !Types.delimiter.match(token) && token.type === _tokens.TokenType.TRUE || token.type === _tokens.TokenType.FALSE
  },

  template: {
    match: token => !Types.delimiter.match(token) && token.type === _tokens.TokenType.TEMPLATE
  },

  delimiter: {
    match: token => _immutable.List.isList(token)
  },

  syntaxTemplate: {
    match: token => Types.delimiter.match(token) && token.get(0).val() === '#`'
  },

  eof: {
    match: token => !Types.delimiter.match(token) && token.type === _tokens.TokenType.EOS
  }
};
const ALL_PHASES = exports.ALL_PHASES = {};

class Syntax {

  constructor(token, oldstx) {
    this.token = token;
    this.bindings = oldstx && oldstx.bindings != null ? oldstx.bindings : new _bindingMap2.default();
    this.scopesets = oldstx && oldstx.scopesets != null ? oldstx.scopesets : {
      all: (0, _immutable.List)(),
      phase: (0, _immutable.Map)()
    };
    Object.freeze(this);
  }
  // token: Token | List<Token>;


  static of(token, stx) {
    return new Syntax(token, stx);
  }

  static from(type, value, stx) {
    if (!Types[type]) {
      throw new Error(type + ' is not a valid type');
    } else if (!Types[type].create) {
      throw new Error('Cannot create a syntax from type ' + type);
    }
    let newstx = Types[type].create(value, stx);
    let slice = getFirstSlice(stx);
    if (slice != null && newstx.token != null) {
      newstx.token.slice = slice;
    }
    return newstx;
  }

  from(type, value) {
    // TODO: this is gross, fix
    let s = Syntax.from(type, value, this);
    if (s instanceof Syntax) {
      return new T.RawSyntax({ value: s });
    }
    return s;
  }

  fromNull() {
    return this.from('null', null);
  }

  fromNumber(value) {
    return this.from('number', value);
  }

  fromString(value) {
    return this.from('string', value);
  }

  fromPunctuator(value) {
    return this.from('punctuator', value);
  }

  fromKeyword(value) {
    return this.from('keyword', value);
  }

  fromIdentifier(value) {
    return this.from('identifier', value);
  }

  fromRegularExpression(value) {
    return this.from('regularExpression', value);
  }

  static fromNull(stx) {
    return Syntax.from('null', null, stx);
  }

  static fromNumber(value, stx) {
    return Syntax.from('number', value, stx);
  }

  static fromString(value, stx) {
    return Syntax.from('string', value, stx);
  }

  static fromPunctuator(value, stx) {
    return Syntax.from('punctuator', value, stx);
  }

  static fromKeyword(value, stx) {
    return Syntax.from('keyword', value, stx);
  }

  static fromIdentifier(value, stx) {
    return Syntax.from('identifier', value, stx);
  }

  static fromRegularExpression(value, stx) {
    return Syntax.from('regularExpression', value, stx);
  }

  // () -> string
  resolve(phase) {
    (0, _errors.assert)(phase != null, 'must provide a phase to resolve');
    let allScopes = this.scopesets.all;
    let stxScopes = this.scopesets.phase.has(phase) ? this.scopesets.phase.get(phase) : (0, _immutable.List)();
    stxScopes = allScopes.concat(stxScopes);
    if (stxScopes.size === 0 || !(this.match('identifier') || this.match('keyword') || this.match('punctuator'))) {
      return this.token.value;
    }
    let scope = stxScopes.last();
    let bindings = this.bindings;
    if (scope) {
      // List<{ scopes: List<Scope>, binding: Symbol }>
      let scopesetBindingList = bindings.get(this);

      if (scopesetBindingList) {
        // { scopes: List<Scope>, binding: Symbol }
        let biggestBindingPair = scopesetBindingList.filter(({ scopes }) => {
          return scopes.isSubset(stxScopes);
        }).sort(sizeDecending);

        // if (
        //   biggestBindingPair.size >= 2 &&
        //   biggestBindingPair.get(0).scopes.size ===
        //     biggestBindingPair.get(1).scopes.size
        // ) {
        //   let debugBase =
        //     '{' + stxScopes.map(s => s.toString()).join(', ') + '}';
        //   let debugAmbigousScopesets = biggestBindingPair
        //     .map(({ scopes }) => {
        //       return '{' + scopes.map(s => s.toString()).join(', ') + '}';
        //     })
        //     .join(', ');
        //   throw new Error(
        //     'Scopeset ' +
        //       debugBase +
        //       ' has ambiguous subsets ' +
        //       debugAmbigousScopesets,
        //   );
        // } else
        if (biggestBindingPair.size !== 0) {
          let bindingStr = biggestBindingPair.get(0).binding.toString();
          if (_ramdaFantasy.Maybe.isJust(biggestBindingPair.get(0).alias)) {
            // null never happens because we just checked if it is a Just
            return biggestBindingPair.get(0).alias.getOrElse(null).resolve(phase);
          }
          return bindingStr;
        }
      }
    }
    return this.token.value;
  }

  val() {
    (0, _errors.assert)(!this.match('delimiter'), 'cannot get the val of a delimiter');
    if (this.match('string')) {
      return this.token.str;
    }
    if (this.match('template')) {
      if (!this.token.items) return this.token.value;
      return this.token.items.map(el => {
        if (typeof el.match === 'function' && el.match('delimiter')) {
          return '${...}';
        }
        return el.slice.text;
      }).join('');
    }
    return this.token.value;
  }

  lineNumber() {
    if (!this.match('delimiter')) {
      return this.token.slice.startLocation.line;
    } else {
      return this.token.get(0).lineNumber();
    }
  }

  setLineNumber(line) {
    let newTok = {};
    if (this.isDelimiter()) {
      newTok = this.token.map(s => s.setLineNumber(line));
    } else {
      for (let key of Object.keys(this.token)) {
        newTok[key] = this.token[key];
      }
      (0, _errors.assert)(newTok.slice && newTok.slice.startLocation, 'all tokens must have line info');
      newTok.slice.startLocation.line = line;
    }
    return new Syntax(newTok, this);
  }

  // () -> List<Syntax>
  // inner() {
  //   assert(this.match("delimiter"), "can only get the inner of a delimiter");
  //   return this.token.slice(1, this.token.size - 1);
  // }

  addScope(scope, bindings, phase, options = { flip: false }) {
    let token = this.match('delimiter') ? this.token.map(s => s.addScope(scope, bindings, phase, options)) : this.token;
    if (this.match('template')) {
      token = _.merge(token, {
        items: token.items.map(it => {
          if (it instanceof Syntax && it.match('delimiter')) {
            return it.addScope(scope, bindings, phase, options);
          }
          return it;
        })
      });
    }
    let oldScopeset;
    if (phase === ALL_PHASES) {
      oldScopeset = this.scopesets.all;
    } else {
      oldScopeset = this.scopesets.phase.has(phase) ? this.scopesets.phase.get(phase) : (0, _immutable.List)();
    }
    let newScopeset;
    if (options.flip) {
      let index = oldScopeset.indexOf(scope);
      if (index !== -1) {
        newScopeset = oldScopeset.remove(index);
      } else {
        newScopeset = oldScopeset.push(scope);
      }
    } else {
      newScopeset = oldScopeset.push(scope);
    }
    let newstx = {
      bindings,
      scopesets: {
        all: this.scopesets.all,
        phase: this.scopesets.phase
      }
    };

    if (phase === ALL_PHASES) {
      newstx.scopesets.all = newScopeset;
    } else {
      newstx.scopesets.phase = newstx.scopesets.phase.set(phase, newScopeset);
    }
    return new Syntax(token, newstx);
  }

  removeScope(scope, phase) {
    let token = this.match('delimiter') ? this.token.map(s => s.removeScope(scope, phase)) : this.token;
    let phaseScopeset = this.scopesets.phase.has(phase) ? this.scopesets.phase.get(phase) : (0, _immutable.List)();
    let allScopeset = this.scopesets.all;
    let newstx = {
      bindings: this.bindings,
      scopesets: {
        all: this.scopesets.all,
        phase: this.scopesets.phase
      }
    };

    let phaseIndex = phaseScopeset.indexOf(scope);
    let allIndex = allScopeset.indexOf(scope);
    if (phaseIndex !== -1) {
      newstx.scopesets.phase = this.scopesets.phase.set(phase, phaseScopeset.remove(phaseIndex));
    } else if (allIndex !== -1) {
      newstx.scopesets.all = allScopeset.remove(allIndex);
    }
    return new Syntax(token, newstx);
  }

  match(type, value) {
    if (!Types[type]) {
      throw new Error(type + ' is an invalid type');
    }
    return Types[type].match(this.token) && (value == null || (value instanceof RegExp ? value.test(this.val()) : this.val() == value));
  }

  isIdentifier(value) {
    return this.match('identifier', value);
  }

  isAssign(value) {
    return this.match('assign', value);
  }

  isBooleanLiteral(value) {
    return this.match('boolean', value);
  }

  isKeyword(value) {
    return this.match('keyword', value);
  }

  isNullLiteral(value) {
    return this.match('null', value);
  }

  isNumericLiteral(value) {
    return this.match('number', value);
  }

  isPunctuator(value) {
    return this.match('punctuator', value);
  }

  isStringLiteral(value) {
    return this.match('string', value);
  }

  isRegularExpression(value) {
    return this.match('regularExpression', value);
  }

  isTemplate(value) {
    return this.match('template', value);
  }

  isDelimiter(value) {
    return this.match('delimiter', value);
  }

  isParens(value) {
    return this.match('parens', value);
  }

  isBraces(value) {
    return this.match('braces', value);
  }

  isBrackets(value) {
    return this.match('brackets', value);
  }

  isSyntaxTemplate(value) {
    return this.match('syntaxTemplate', value);
  }

  isEOF(value) {
    return this.match('eof', value);
  }

  toString() {
    if (this.match('delimiter')) {
      return this.token.map(s => s.toString()).join(' ');
    }
    if (this.match('string')) {
      return `'${this.token.str}'`;
    }
    if (this.match('template')) {
      return this.val();
    }
    return this.token.value;
  }
}
exports.default = Syntax;
