// @sweet-js/core/dist/transforms.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompiletimeTransform = exports.VarBindingTransform = exports.ModuleNamespaceTransform = exports.ReturnStatementTransform = exports.OperatorDeclTransform = exports.SyntaxDeclTransform = exports.SyntaxrecDeclTransform = exports.DebuggerTransform = exports.DefaultTransform = exports.ClassTransform = exports.ThisTransform = exports.YieldTransform = exports.SuperTransform = exports.ExportTransform = exports.ImportTransform = exports.WithTransform = exports.DoTransform = exports.ContinueTransform = exports.BreakTransform = exports.SwitchTransform = exports.ForTransform = exports.IfTransform = exports.WhileTransform = exports.TryTransform = exports.ConstDeclTransform = exports.LetDeclTransform = exports.ThrowTransform = exports.NewTransform = exports.VariableDeclTransform = exports.FunctionDeclTransform = undefined;

var _sweetModule = require('./sweet-module');

var _sweetModule2 = _interopRequireDefault(_sweetModule);

var _syntax = require('./syntax');

var _syntax2 = _interopRequireDefault(_syntax);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class FunctionDeclTransform {}
exports.FunctionDeclTransform = FunctionDeclTransform;
class VariableDeclTransform {}
exports.VariableDeclTransform = VariableDeclTransform;
class NewTransform {}
exports.NewTransform = NewTransform;
class ThrowTransform {}
exports.ThrowTransform = ThrowTransform;
class LetDeclTransform {}
exports.LetDeclTransform = LetDeclTransform;
class ConstDeclTransform {}
exports.ConstDeclTransform = ConstDeclTransform;
class TryTransform {}
exports.TryTransform = TryTransform;
class WhileTransform {}
exports.WhileTransform = WhileTransform;
class IfTransform {}
exports.IfTransform = IfTransform;
class ForTransform {}
exports.ForTransform = ForTransform;
class SwitchTransform {}
exports.SwitchTransform = SwitchTransform;
class BreakTransform {}
exports.BreakTransform = BreakTransform;
class ContinueTransform {}
exports.ContinueTransform = ContinueTransform;
class DoTransform {}
exports.DoTransform = DoTransform;
class WithTransform {}
exports.WithTransform = WithTransform;
class ImportTransform {}
exports.ImportTransform = ImportTransform;
class ExportTransform {}
exports.ExportTransform = ExportTransform;
class SuperTransform {}
exports.SuperTransform = SuperTransform;
class YieldTransform {}
exports.YieldTransform = YieldTransform;
class ThisTransform {}
exports.ThisTransform = ThisTransform;
class ClassTransform {}
exports.ClassTransform = ClassTransform;
class DefaultTransform {}
exports.DefaultTransform = DefaultTransform;
class DebuggerTransform {}
exports.DebuggerTransform = DebuggerTransform;
class SyntaxrecDeclTransform {}
exports.SyntaxrecDeclTransform = SyntaxrecDeclTransform;
class SyntaxDeclTransform {}
exports.SyntaxDeclTransform = SyntaxDeclTransform;
class OperatorDeclTransform {}
exports.OperatorDeclTransform = OperatorDeclTransform;
class ReturnStatementTransform {}
exports.ReturnStatementTransform = ReturnStatementTransform;
class ModuleNamespaceTransform {

  constructor(namespace, mod) {
    this.namespace = namespace;
    this.mod = mod;
  }
}
exports.ModuleNamespaceTransform = ModuleNamespaceTransform;
class VarBindingTransform {

  constructor(id) {
    this.id = id;
  }
}
exports.VarBindingTransform = VarBindingTransform;
class CompiletimeTransform {

  constructor(value) {
    this.value = value;
  }
}
exports.CompiletimeTransform = CompiletimeTransform;
