// readtable/dist/reader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setCurrentReadtable = setCurrentReadtable;
exports.getCurrentReadtable = getCurrentReadtable;

var _charStream = require('./char-stream');

var _readtable = require('./readtable');

const defaultDispatchKey = '#';
let dispatching = false;

let currentReadtable = _readtable.EmptyReadtable.extend({
  key: defaultDispatchKey,
  mode: 'non-terminating',
  action: function readDispatchChar(stream, ...rest) {
    stream.readString();
    dispatching = true;
    return this.read(stream, ...rest, defaultDispatchKey);
  }
});

class Reader {
  read(stream, ...rest) {
    let key = stream.peek();
    if (!(0, _charStream.isEOS)(key)) {
      const entry = currentReadtable.getMapping(key);
      const action = dispatching ? (dispatching = false, entry.dispatchAction) : entry.action;

      return action.call(this, stream, ...rest);
    }
    throw Error('Unexpected end of input');
  }
}

exports.default = Reader;
function setCurrentReadtable(readtable) {
  currentReadtable = readtable;
}

function getCurrentReadtable() {
  return currentReadtable;
}
