// sweet.js/dist/sweet.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expand = expand;
exports.parse = parse;
exports.compile = compile;

var _shiftReader = require("./shift-reader");

var _shiftReader2 = _interopRequireDefault(_shiftReader);

var _expander = require("./expander");

var _expander2 = _interopRequireDefault(_expander);

var _immutable = require("immutable");

var _syntax = require("./syntax");

var _syntax2 = _interopRequireDefault(_syntax);

var _env = require("./env");

var _env2 = _interopRequireDefault(_env);

var _shiftReducer = require("shift-reducer");

var _shiftReducer2 = _interopRequireDefault(_shiftReducer);

var _parseReducer = require("./parse-reducer");

var _parseReducer2 = _interopRequireDefault(_parseReducer);

var _shiftCodegen = require("shift-codegen");

var _shiftCodegen2 = _interopRequireDefault(_shiftCodegen);

var _scope = require("./scope");

var _bindingMap = require("./binding-map.js");

var _bindingMap2 = _interopRequireDefault(_bindingMap);

var _terms = require("./terms");

var _terms2 = _interopRequireDefault(_terms);

var _symbol = require("./symbol");

var _modules = require("./modules");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function expand(source) {
  var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

  var reader = new _shiftReader2.default(source);
  var stxl = reader.read();
  var scope = (0, _scope.freshScope)('top');
  var bindings = new _bindingMap2.default();
  var expander = new _expander2.default({
    env: new _env2.default(),
    store: new _env2.default(),
    bindings: bindings,
    cwd: options.cwd,
    modules: new _modules.Modules(),
    currentScope: [scope],
    transform: options.transform ? options.transform : function (x) {
      return { code: x };
    },
    moduleResolver: options.moduleResolver,
    moduleLoader: options.moduleLoader
  });
  var exStxl = expander.expand(stxl.map(function (s) {
    return s.addScope(scope, bindings);
  }));
  return new _terms2.default("Module", {
    directives: (0, _immutable.List)(),
    items: exStxl
  });
}

function parse(source) {
  var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

  return (0, _shiftReducer2.default)(new _parseReducer2.default(), expand(source, options));
}

function compile(source, opt) {
  var ast = parse(source, opt);
  var gen = (0, _shiftCodegen2.default)(ast);
  return opt.transform ? opt.transform(gen) : { code: gen };
}
