// autodiff/lib-esm/series/autodiff-context.js
import { Pool } from './pool';
import { Series } from './series';
import { Series2D } from './series2d';
function newSeries() {
    return new Series();
}
function clearSeries(s) {
    for (var i = 0; i < s.coefficients.length; i++) {
        s.coefficients[i] = 0;
    }
}
function copySeries(to, from) {
    for (var i = 0; i < to.coefficients.length; i++) {
        to.coefficients[i] = from.coefficients[i];
    }
}
function newSeries2D() {
    return new Series2D();
}
function clearSeries2D(s) {
    for (var i = 0; i < s.coefficients.length; i++) {
        s.coefficients[i] = 0;
    }
}
function copySeries2D(to, from) {
    for (var i = 0; i < to.coefficients.length; i++) {
        to.coefficients[i] = from.coefficients[i];
    }
}
var AutodiffContext = (function () {
    function AutodiffContext(opts) {
        this._numDerivativesToCompute = 3;
        this.seriesPool = new Pool(newSeries, clearSeries, copySeries);
        this.series2DPool = new Pool(newSeries2D, clearSeries2D, copySeries2D);
        this._numDerivativesToCompute = opts.numDerivativesToCompute;
    }
    AutodiffContext.prototype.setNumberOfDerivativesToCompute = function (n) {
        if (n !== this._numDerivativesToCompute) {
            this.seriesPool.forgetFreeElements();
            this.series2DPool.forgetFreeElements();
        }
        this._numDerivativesToCompute = n;
    };
    AutodiffContext.prototype.numDerivativesToCompute = function () {
        return this._numDerivativesToCompute;
    };
    return AutodiffContext;
}());
export { AutodiffContext };
export var defaultContext = new AutodiffContext({
    numDerivativesToCompute: 5
});
