// @sweet-js/core/dist/reader/char-stream.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isEOS = isEOS;


const sourceInfo = new WeakMap();
class CharStream {
  constructor(source, filename = '') {
    sourceInfo.set(this, {
      source,
      filename,
      position: 0
    });
  }

  get sourceInfo() {
    // $FlowFixMe: decide on how to handle possible nullability
    const { filename, position /*, source*/ } = sourceInfo.get(this);
    return { filename, position /*, source*/ };
  }

  // returns the Unicode character charsToSkip ahead.
  peek(charsToSkip = 0) {
    // $FlowFixMe: decide on how to handle possible nullability
    const { source, position } = sourceInfo.get(this);
    if (position + charsToSkip >= source.length) return '';
    return source[position + charsToSkip];
  }

  // returns a string containing the next numChars characters.
  readString(numChars = 1) {
    const info = sourceInfo.get(this);
    // $FlowFixMe: decide on how to handle possible nullability
    const { source, position } = info;
    const str = source.slice(position, position + numChars);
    // $FlowFixMe: decide on how to handle possible nullability
    info.position += str.length;
    return str;
  }

  getSlice(start) {
    // $FlowFixMe: decide on how to handle possible nullability
    const { source, position } = sourceInfo.get(this);
    return source.slice(start, position);
  }
}

exports.default = CharStream;
function isEOS(char) {
  return char === '';
}
