// @sweet-js/core/dist/reader/read-dispatch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readSyntaxTemplate = readSyntaxTemplate;

var _readtable = require('readtable');

var _immutable = require('immutable');

var _tokens = require('../tokens');

const backtickEntry = {
  key: '`',
  mode: 'terminating',
  action: function readBacktick(stream, prefix, e) {
    if (prefix.isEmpty()) {
      return {
        type: _tokens.TokenType.LSYNTAX,
        value: stream.readString()
      };
    }

    return {
      type: _tokens.TokenType.RSYNTAX,
      value: stream.readString()
    };
  }
};

function readSyntaxTemplate(stream, prefix, exprAllowed, dispatchChar) {
  // return read('syntaxTemplate').first().token;
  // TODO: Can we simply tack 'syntaxTemplate' on the front and process it as a
  //       syntax macro?
  const prevTable = (0, _readtable.getCurrentReadtable)();
  (0, _readtable.setCurrentReadtable)(prevTable.extend(backtickEntry));

  const result = this.readUntil('`', stream, _immutable.List.of(updateSyntax(dispatchChar, this.readToken(stream, (0, _immutable.List)(), exprAllowed))), exprAllowed);

  (0, _readtable.setCurrentReadtable)(prevTable);
  return result;
}

function updateSyntax(prefix, token) {
  token.value = prefix + token.value;
  token.slice.text = prefix + token.slice.text;
  token.slice.start -= 1;
  token.slice.startLocation.position -= 1;
  return token;
}
